fendermint/testing/contract-test/tests/upgrades.rs

```diff
@@ -0,0 +1,281 @@
+#[tokio::test]
+async fn test_applying_upgrades() {
+    use bytes::Bytes;
+    use fendermint_rpc::message::{GasParams, MessageFactory};
+    use lazy_static::lazy_static;
+
+    lazy_static! {
+        /// Default gas params based on the testkit.
+        static ref GAS_PARAMS: GasParams = GasParams {
+            gas_limit: 10_000_000_000,
+            gas_fee_cap: TokenAmount::default(),
+            gas_premium: TokenAmount::default(),
+        };
+    }
+
+    // this is the contract we want to deploy
+    const CONTRACT_HEX: &str = include_str!("../../contracts/SimpleCoin.bin");
+    // generate type safe bindings in rust to this contract
+    abigen!(SimpleCoin, "../contracts/SimpleCoin.abi");
+    // once we deploy this contract, this is the address we expect the contract to be deployed to
+    const CONTRACT_ADDRESS: &str = "f410fnz5jdky3zzcj6pejqkomkggw72pcuvkpihz2rwa";
+    // the amount we want to send to the contract
+    const SEND_BALANCE_AMOUNT: u64 = 1000;
+
+    let mut upgrade_scheduler = UpgradeScheduler::new();
+    upgrade_scheduler
+        .add(
+            Upgrade::new("mychain".to_string(), 1, |state| {
+                println!(
+                    "[Upgrade at height {}] Deploy simple contract",
+                    state.block_height()
+                );
+
+                // create a message for deploying the contract
+                let mut mf = MessageFactory::new_secp256k1(my_secret_key(), 1, state.chain_id());
+                let message = match mf
+                    .fevm_create(
+                        Bytes::from(
+                            hex::decode(CONTRACT_HEX)
+                                .context("error parsing contract")
+                                .unwrap(),
+                        ),
+                        Bytes::default(),
+                        TokenAmount::default(),
+                        GAS_PARAMS.clone(),
+                    )
+                    .unwrap()
+                {
+                    ChainMessage::Signed(signed) => signed.into_message(),
+                    _ => panic!("unexpected message type"),
+                };
+
+                // execute the message
+                let (res, _) = state.execute_implicit(message).unwrap();
+                assert!(
+                    res.msg_receipt.exit_code.is_success(),
+                    "{:?}",
+                    res.failure_info
+                );
+
+                // parse the message receipt data and make sure the contract was deployed to the expected address
+                let res = fvm_ipld_encoding::from_slice::<eam::CreateReturn>(
+                    &res.msg_receipt.return_data,
+                )
+                .unwrap();
+                assert_eq!(
+                    res.delegated_address(),
+                    Address::from_str(CONTRACT_ADDRESS).unwrap()
+                );
+
+                // lets upgrade the app version as well
+                let app_version = 1;
```

Comment:
Should the app version be an actual _property_ of the upgrade, like the block height itself? That way we can validate that successive upgrades are incrementing the app version, without running them. What's the argument for making it a computed value?

Reply:
I think that makes sense as it pulls the version up to the top making it also visible that its indeed upgrading the app version (without requiring to parse the `MigrationFunc` code)